### PR TITLE
Issue #615: Introduce store level statistics

### PR DESCRIPTION
**Pull request by ljacomet**
* Access stat for store mutations
* Tier efficiency stat for all tier specific operations, trying to indicate how well the tier is working

### CODE REVIEW

File: core/src/main/java/org/ehcache/spi/cache/tiering/LowerCachingTier.java [hunk 1]
```diff
@@ -38,7 +38,7 @@
    * @return the value holder, or null.
    * @throws CacheAccessException
    */
-  Store.ValueHolder<V> getOrComputeIfAbsent(K key, Function<K, Store.ValueHolder<V>> source) throws CacheAccessException;
+  Store.ValueHolder<V> installMapping(K key, Function<K, Store.ValueHolder<V>> source) throws CacheAccessException;
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/spi/cache/tiering/LowerCachingTier.java]: Why are we renaming this method in this PR?  And if it's necessary could we do the rename in it's own commit?

File: core/src/main/java/org/ehcache/statistics/StoreOperationOutcomes.java [hunk 2]
```diff
@@ -22,38 +22,24 @@
...
-  };
-
-  /**
-   * The Enum RemoveOutcome.
-   */
-  enum RemoveOutcome {
-    /** success. */
-    SUCCESS
-  };
+  enum TierEfficiencyOutcome implements StoreOperationOutcomes {
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/statistics/StoreOperationOutcomes.java]: Efficiency seems like a weird name.  Efficiency would be something derived from this raw data right?  TierUtilityOutcome?  TierHelpfullnessOutcome?

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java [hunk 3]
```diff
@@ -124,6 +124,8 @@ public void cacheConfigurationChange(CacheConfigurationChangeEvent event) {
     }
   };
 
+  private final OperationObserver<StoreOperationOutcomes.AccessOutcome> accessObserver = operation(StoreOperationOutcomes.AccessOutcome.class).named("access").of(this).tag("onheap-store").build();
+  private final OperationObserver<StoreOperationOutcomes.TierEfficiencyOutcome> tierEfficiencyObserver = operation(StoreOperationOutcomes.TierEfficiencyOutcome.class).named("faulting").of(this).tag("onheap-store").build();
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: Why is the TierEfficiencyOutcome which we have tried to abstract from the implementation called "faulting"?

File: core/src/main/java/org/ehcache/spi/cache/tiering/LowerCachingTier.java (hunk 1 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/spi/cache/tiering/LowerCachingTier.java]: It is a `LowerCachingTier` method that has the exact same name as a `CachingTier` method but a completely different contract. Since I was touching code in the area I could not resist the rename that really should have been part of the PR introducing this interface. I will isolate in its own commit and reference that issue instead.

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java [hunk 4]
```diff
@@ -174,7 +176,8 @@ public OnHeapStore(final Configuration<K, V> config, TimeSource timeSource, Copi
   }
   
   private OnHeapValueHolder<V> internalGet(final K key, final boolean updateAccess) throws CacheAccessException {
-    return map.computeIfPresent(key, new BiFunction<K, OnHeapValueHolder<V>, OnHeapValueHolder<V>>() {
+    accessObserver.begin();
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: I'm ignorant of lots of the details here... but it seems weird that we're calling the `accessObserver` here, but not calling the `tierEfficiencyObserver`.

File: impl/src/main/java/org/ehcache/internal/store/offheap/AbstractOffHeapStore.java [hunk 5]
```diff
@@ -149,15 +146,17 @@ public void put(final K key, final V value) throws CacheAccessException {
             }
           }
         }, false);
-        return;
+        break;
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/internal/store/offheap/AbstractOffHeapStore.java]: Could we not just put the accessObserver call before the return?

File: impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java [hunk 6]
```diff
@@ -581,14 +639,16 @@ public void testComputeIfAbsentExisting() throws Exception {
     OnHeapStore<String, String> store = newStore();
 
     store.put("key", "value");
-    ValueHolder<String> newValue = store.computeIfAbsent("key", new Function<String, String>() {
+    ValueHolder<String> newValue =  store.computeIfAbsent("key", new Function<String, String>() {
```
Comment (chrisdennis) [inline impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java]: no change here

File: impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java [hunk 7]
```diff
@@ -618,7 +679,7 @@ public String apply(String mappedKey) {
           throw RUNTIME_EXCEPTION;
         }
       });
-      throw new AssertionError();
+      fail("Expected exception");
```
Comment (chrisdennis) [inline impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java]: why the change?

File: impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java [hunk 8]
```diff
@@ -852,7 +904,7 @@ public void testGetOrComputeIfAbsentContention() throws InterruptedException {
 
     final OnHeapStore<Long, String> store = newStore();
 
-    int threads = 10;
+    int threads = 5;
```
Comment (chrisdennis) [inline impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java]: why?

File: core/src/main/java/org/ehcache/statistics/StoreOperationOutcomes.java (hunk 2 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/statistics/StoreOperationOutcomes.java]: Indeed - will use Helpfulness as it relates to the outcomes.

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java (hunk 3 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: Good catch!

File: impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java (hunk 7 above)
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java]: To add a message in case this ever fails, applying boy scout rule while navigating (and changing this code)

File: impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java (hunk 6 above)
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java]: Will revert

File: impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java (hunk 8 above)
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/internal/store/heap/BaseOnHeapStoreTest.java]: Will revert

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java (hunk 4 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: When the `OnHeapStore` acts as anything but an authority, this method is never called, so no need to involve the tier specific stat.

File: impl/src/main/java/org/ehcache/internal/store/offheap/AbstractOffHeapStore.java (hunk 5 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/internal/store/offheap/AbstractOffHeapStore.java]: Indeed, will move back in there and take care of the other `break` the same.

**Comment by ljacomet** [comment]
Having second thoughts - will open an other PR with a different impl for this.
